Sources/Apollo/DataLoader.swift

```diff
@@ -1,52 +1,42 @@
 import Dispatch
 
 final class DataLoader<Key: Hashable, Value> {
-  public typealias BatchLoad = ([Key]) -> Promise<[Value]>
-  typealias Load = (key: Key, fulfill: (Value) -> Void, reject: (Error) -> Void)
-
+  public typealias BatchLoad = (Set<Key>) throws -> [Key: Value]
   private var batchLoad: BatchLoad
 
-  private var cache: [Key: Promise<Value>] = [:]
-  private var loads: [Load] = []
+  private var cache: [Key: Result<Value?, Error>] = [:]
+  private var pendingLoads: Set<Key> = []
 
   public init(_ batchLoad: @escaping BatchLoad) {
     self.batchLoad = batchLoad
   }
 
-  subscript(key: Key) -> Promise<Value> {
-    if let promise = cache[key] {
-      return promise
-    }
-
-    let promise = Promise<Value> { fulfill, reject in
-      enqueue(load: (key, fulfill, reject))
+  subscript(key: Key) -> PossiblyDeferred<Value?> {
+    if let cachedResult = cache[key] {
+      return .immediate(cachedResult)
     }
+    
+    pendingLoads.insert(key)
 
-    cache[key] = promise
-
-    return promise
+    return .deferred { try self.load(key) }
   }
-
-  private func enqueue(load: Load) {
-    self.loads.append(load)
-  }
-
-  func dispatch() {
-    let loads = self.loads
-
-    if loads.isEmpty { return }
-
-    self.loads = []
-
-    let keys = loads.map { $0.key }
-
-    self.batchLoad(keys).catch { error in
-      loads.forEach { $0.reject(error) }
-    }.andThen { values in
-      for (load, value) in zip(loads, values) {
-        load.fulfill(value)
-      }
+  
+  func load(_ key: Key) throws -> Value? {
+    if let cachedResult = cache[key] {
+      return try cachedResult.get()
+    }
+    
+    assert(pendingLoads.contains(key))
```

Comment:
Just a check to make sure the API isn't being misused. It should have no effect in release builds. Now that I think of it though, we should also make that method private.

Reply:
Yeah I think making it private would probably be a way better gate against API misuse if we can swing it

Reply:
Now that this is private do we still need the assertion here?

Reply:
It seems like a good safety net to have. I would like to start using more assertions actually, to catch violations of expectations early in case we make changes later.

Reply:
Yeah...we should have a longer chat about that because if they are hit in weird situations from the client, they can be super-confusing to the caller. For now, I'm good leaving this be